framework/src/meshmodifiers/ElementSubdomainModifierBase.C

```diff
@@ -0,0 +1,632 @@
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ElementSubdomainModifierBase.h"
+#include "DisplacedProblem.h"
+
+#include "libmesh/parallel_algebra.h"
+#include "libmesh/parallel.h"
+#include "libmesh/dof_map.h"
+#include "libmesh/remote_elem.h"
+#include "libmesh/parallel_ghost_sync.h"
+#include "libmesh/petsc_vector.h"
+
+InputParameters
+ElementSubdomainModifierBase::validParams()
+{
+  InputParameters params = ElementUserObject::validParams();
+
+  // ESMs only operate on the undisplaced mesh to gather subdomain and sideset information. This
+  // information is used to modify both the undisplaced and the displaced meshes. It is the
+  // developer's responsibility to make sure the element IDs and sideset info are consistent across
+  // both meshes.
+  params.set<bool>("use_displaced_mesh") = false;
+  params.suppressParameter<bool>("use_displaced_mesh");
+
+  params.addParam<std::vector<BoundaryName>>(
+      "moving_boundaries",
+      {},
+      "Moving boundaries between subdomains. These boundaries (both sidesets and nodesets) will be "
+      "updated as elements change their subdomain. The corresponding subdomains of each moving "
+      "boundary shall be specified using the parameter 'moving_boundary_subdomain_pairs'. If one "
+      "boundary and multiple subdomain pairs are specified, then it is assumed that the pairs all "
+      "apply to the boundary. A boundary will be created on the mesh if it does not already "
+      "exist.");
+  params.addParam<std::vector<std::vector<SubdomainName>>>(
+      "moving_boundary_subdomain_pairs",
+      {},
+      "The subdomain pairs associated with each moving boundary. For each pair of subdomains, only "
+      "the element side from the first subdomain will be added to the moving boundary, i.e., the "
+      "side normal is pointing from the first subdomain to the second subdomain. The pairs shall "
+      "be delimited by ';'. If a pair only has one subdomain, the moving boundary is associated "
+      "with the subdomain's external boundary, i.e., when the elements have no neighboring "
+      "elements.");
+
+  params.addParam<std::vector<SubdomainName>>(
+      "reinitialize_subdomains",
+      {"ANY_BLOCK_ID"},
+      "By default, any element which changes subdomain is reinitialized. If a list of subdomains "
+      "(IDs or names) is set, then only elements which change to a subdomain in the list will be "
+      "reinitialized. If an empty list is set, then no elements will be reinitialized.");
+  params.addParam<bool>(
+      "previous_subdomain_reinitialized",
+      true,
+      "If set to false, only elements which change from subdomains not listed in "
+      "'reinitialize_subdomains', to ones that are listed, are reinitialized. "
+      "This parameter must be set with a non-empty list in 'reinitialize_subdomains'.");
```

Comment:
It is a little counter intuitive here, I guess the idea is to prevent reinitialize elements multiple times if it is in the `reinitialize_subdomains`? Then maybe call this as something like "allow_duplicate_reinit"?

Reply:
Hello, sorry, this parameter name is a bit confusing and I'm open to suggestions for it. `previous_subdomain_reinitialized` is an optional parameter that gives greater control over which elements are reinitialized. In addition to checking whether an element's new subdomain is in the `reinitialize_subdomains`, the element's old subdomain will also be checked. If `previous_subdomain_reinitialized` is set to false, then that element will only be reinitialized if its old subdomain isn't in `reinitialize_subdomains` (see test file `reinitialization_from_into`). This is useful, for example, when you want to reinitialize elements that move from inactive to active subdomains, but ignore any elements that change from one active subdomain to another active subdomain.

Reply:
Thanks for your detailed explanation! Now I get your point. I think it may be okay to keep the name as you have, but maybe add a bit more explanation about the default behavior when `previous_subdomain_reinitialized =true`?

Reply:
I have updated the docstring for the parameter to hopefully be clearer (I also changed the parameter's name to `old_subdomain_reinitialized`, mainly to make it shorter)

Reply:
Thanks! Looks good to me!